File: apps/web/lib/verse_diff.ts
/** 对照译本措辞差异（轻量 LCS，按节计算，可缓存）。 */

export type DiffSide = 'main' | 'parallel';

export type DiffSpan = {
  start: number;
  end: number;
  side: DiffSide;
};

export type VerseDiffResult = {
  main: DiffSpan[];
  parallel: DiffSpan[];
  /** 差异过多时降级为整节提示 */
  heavy: boolean;
};

const MAX_SPANS = 40;
const PUNCT_RE = /[\s\u3000，。！？、；：""''（）【】《》…—\-.,!?;:'"()[\]{}]/g;

function normalizeChar(ch: string): string | null {
  if (PUNCT_RE.test(ch)) return null;
  return ch;
}

function tokenize(text: string): { chars: string[]; indexMap: number[] } {
  const chars: string[] = [];
  const indexMap: number[] = [];
  const arr = [...(text || '')];
  for (let i = 0; i < arr.length; i += 1) {
    const n = normalizeChar(arr[i]!);
    if (n == null) continue;
    chars.push(n);
    indexMap.push(i);
  }
  return { chars, indexMap };
}

/** 简单 LCS 标记：不在 LCS 中的 token 视为差异。 */
function lcsMask(a: string[], b: string[]): { aKeep: boolean[]; bKeep: boolean[] } {
  const n = a.length;
  const m = b.length;
  // 过长节：跳过精细 diff，避免 O(nm) 卡顿
  if (n * m > 12_000) {
    return {
      aKeep: a.map(() => true),
      bKeep: b.map(() => true),
    };
  }
  const dp: Uint16Array[] = Array.from({ length: n + 1 }, () => new Uint16Array(m + 1));
  for (let i = 1; i <= n; i += 1) {
    const ai = a[i - 1];
    const row = dp[i]!;
    const prev = dp[i - 1]!;
    for (let j = 1; j <= m; j += 1) {
      row[j] = ai === b[j - 1] ? (prev[j - 1]! + 1) : Math.max(prev[j]!, row[j - 1]!);
    }
  }
  const aKeep = Array(n).fill(false);
  const bKeep = Array(m).fill(false);
  let i = n;
  let j = m;
  while (i > 0 && j > 0) {
    if (a[i - 1] === b[j - 1]) {
      aKeep[i - 1] = true;
      bKeep[j - 1] = true;
      i -= 1;
      j -= 1;
    } else if ((dp[i - 1]![j] ?? 0) >= (dp[i]![j - 1] ?? 0)) {
      i -= 1;
    } else {
      j -= 1;
    }
  }
  return { aKeep, bKeep };
}

function maskToSpans(
  keep: boolean[],
  indexMap: number[],
  side: DiffSide,
): DiffSpan[] {
  const spans: DiffSpan[] = [];
  let i = 0;
  while (i < keep.length) {
    if (keep[i]) {
      i += 1;
      continue;
    }
    const startTok = i;
    while (i < keep.length && !keep[i]) i += 1;
    const endTok = i - 1;
    const start = indexMap[startTok]!;
    const end = indexMap[endTok]! + 1;
    spans.push({ start, end, side });
  }
  return spans;
}

export function diffVerseTexts(mainText: string, parallelText: string): VerseDiffResult {
  const a = tokenize(mainText);
  const b = tokenize(parallelText);
  if (!a.chars.length || !b.chars.length) {
    return { main: [], parallel: [], heavy: false };
  }
  if (a.chars.length * b.chars.length > 12_000) {
    return { main: [], parallel: [], heavy: true };
  }
  const { aKeep, bKeep } = lcsMask(a.chars, b.chars);
  let main = maskToSpans(aKeep, a.indexMap, 'main');
  let parallel = maskToSpans(bKeep, b.indexMap, 'parallel');
  const heavy = main.length + parallel.length > MAX_SPANS;
  if (heavy) {
    main = [];
    parallel = [];
  }
  return { main, parallel, heavy };
}

const memCache = new Map<string, VerseDiffResult>();
const MEM_MAX = 240;

export function cachedVerseDiff(
  key: string,
  mainText: string,
  parallelText: string,
): VerseDiffResult {
  const hit = memCache.get(key);
  if (hit) return hit;
  const result = diffVerseTexts(mainText, parallelText);
  if (memCache.size >= MEM_MAX) {
    const first = memCache.keys().next().value;
    if (first != null) memCache.delete(first);
  }
  memCache.set(key, result);
  return result;
}

export function clearVerseDiffCache() {
  memCache.clear();
}

/** 同文种粗判：双方都含大量 CJK 或都不含。 */
export function sameScriptRoughly(a: string, b: string): boolean {
  const cjk = (t: string) => (/[\u4e00-\u9fff]/.test(t) ? 1 : 0);
  return cjk(a) === cjk(b);
}

export function renderTextWithDiffSpans(
  text: string,
  spans: DiffSpan[],
): { key: string; text: string; diff: boolean }[] {
  if (!spans.length) return [{ key: 'all', text, diff: false }];
  const sorted = [...spans].sort((x, y) => x.start - y.start);
  const parts: { key: string; text: string; diff: boolean }[] = [];
  let cursor = 0;
  sorted.forEach((s, idx) => {
    if (s.start > cursor) {
      parts.push({ key: `t${idx}`, text: text.slice(cursor, s.start), diff: false });
    }
    parts.push({ key: `d${idx}`, text: text.slice(s.start, s.end), diff: true });
    cursor = s.end;
  });
  if (cursor < text.length) {
    parts.push({ key: 'tail', text: text.slice(cursor), diff: false });
  }
  return parts.filter((p) => p.text);
}

File: apps/mobile/lib/features/bible/verse_diff.dart
/// 对照译本措辞差异（轻量 LCS，对齐 Web `verse_diff.ts`）。
library;

enum DiffSide { main, parallel }

class DiffSpan {
  const DiffSpan({required this.start, required this.end, required this.side});
  final int start;
  final int end;
  final DiffSide side;
}

class VerseDiffResult {
  const VerseDiffResult({
    required this.main,
    required this.parallel,
    required this.heavy,
  });
  final List<DiffSpan> main;
  final List<DiffSpan> parallel;
  final bool heavy;
}

const _maxSpans = 40;
final _punctRe = RegExp(
  r'''[\s\u3000，。！？、；：""''（）【】《》…—\-\.,!?;:'"\(\)\[\]{}]''',
);

String? _normalizeChar(String ch) {
  if (_punctRe.hasMatch(ch)) return null;
  return ch;
}

({List<String> chars, List<int> indexMap}) _tokenize(String text) {
  final chars = <String>[];
  final indexMap = <int>[];
  final runes = text.runes.toList();
  var i = 0;
  for (final r in runes) {
    final ch = String.fromCharCode(r);
    final n = _normalizeChar(ch);
    if (n != null) {
      chars.add(n);
      indexMap.add(i);
    }
    i += ch.length;
  }
  return (chars: chars, indexMap: indexMap);
}

({List<bool> aKeep, List<bool> bKeep}) _lcsMask(List<String> a, List<String> b) {
  final n = a.length;
  final m = b.length;
  if (n * m > 12000) {
    return (
      aKeep: List.filled(n, true),
      bKeep: List.filled(m, true),
    );
  }
  final dp = List.generate(n + 1, (_) => List<int>.filled(m + 1, 0));
  for (var i = 1; i <= n; i++) {
    for (var j = 1; j <= m; j++) {
      dp[i][j] = a[i - 1] == b[j - 1]
          ? dp[i - 1][j - 1] + 1
          : (dp[i - 1][j] > dp[i][j - 1] ? dp[i - 1][j] : dp[i][j - 1]);
    }
  }
  final aKeep = List.filled(n, false);
  final bKeep = List.filled(m, false);
  var i = n;
  var j = m;
  while (i > 0 && j > 0) {
    if (a[i - 1] == b[j - 1]) {
      aKeep[i - 1] = true;
      bKeep[j - 1] = true;
      i--;
      j--;
    } else if (dp[i - 1][j] >= dp[i][j - 1]) {
      i--;
    } else {
      j--;
    }
  }
  return (aKeep: aKeep, bKeep: bKeep);
}

List<DiffSpan> _maskToSpans(
  List<bool> keep,
  List<int> indexMap,
  DiffSide side,
) {
  final spans = <DiffSpan>[];
  var i = 0;
  while (i < keep.length) {
    if (keep[i]) {
      i++;
      continue;
    }
    final startTok = i;
    while (i < keep.length && !keep[i]) {
      i++;
    }
    final endTok = i - 1;
    spans.add(DiffSpan(
      start: indexMap[startTok],
      end: indexMap[endTok] + 1,
      side: side,
    ));
  }
  return spans;
}

VerseDiffResult diffVerseTexts(String mainText, String parallelText) {
  final a = _tokenize(mainText);
  final b = _tokenize(parallelText);
  if (a.chars.isEmpty || b.chars.isEmpty) {
    return const VerseDiffResult(main: [], parallel: [], heavy: false);
  }
  if (a.chars.length * b.chars.length > 12000) {
    return const VerseDiffResult(main: [], parallel: [], heavy: true);
  }
  final mask = _lcsMask(a.chars, b.chars);
  var main = _maskToSpans(mask.aKeep, a.indexMap, DiffSide.main);
  var parallel = _maskToSpans(mask.bKeep, b.indexMap, DiffSide.parallel);
  final heavy = main.length + parallel.length > _maxSpans;
  if (heavy) {
    main = [];
    parallel = [];
  }
  return VerseDiffResult(main: main, parallel: parallel, heavy: heavy);
}

final _memCache = <String, VerseDiffResult>{};
const _memMax = 240;

VerseDiffResult cachedVerseDiff(
  String key,
  String mainText,
  String parallelText,
) {
  final hit = _memCache[key];
  if (hit != null) return hit;
  final result = diffVerseTexts(mainText, parallelText);
  if (_memCache.length >= _memMax) {
    _memCache.remove(_memCache.keys.first);
  }
  _memCache[key] = result;
  return result;
}

bool sameScriptRoughly(String a, String b) {
  final cjkA = RegExp(r'[\u4e00-\u9fff]').hasMatch(a);
  final cjkB = RegExp(r'[\u4e00-\u9fff]').hasMatch(b);
  return cjkA == cjkB;
}

List<({String text, bool diff})> renderTextWithDiffSpans(
  String text,
  List<DiffSpan> spans,
) {
  if (spans.isEmpty) return [(text: text, diff: false)];
  final sorted = [...spans]..sort((x, y) => x.start.compareTo(y.start));
  final parts = <({String text, bool diff})>[];
  var cursor = 0;
  for (var idx = 0; idx < sorted.length; idx++) {
    final s = sorted[idx];
    if (s.start > cursor) {
      parts.add((text: text.substring(cursor, s.start), diff: false));
    }
    parts.add((
      text: text.substring(s.start, s.end.clamp(0, text.length)),
      diff: true,
    ));
    cursor = s.end;
  }
  if (cursor < text.length) {
    parts.add((text: text.substring(cursor), diff: false));
  }
  return parts.where((p) => p.text.isNotEmpty).toList();
}
